### PR TITLE
BVAL-508 Add the API for value extraction

### PR DESCRIPTION
**Pull request by gsmet**
@gunnarmorling I went for valueextraction for the package name as it's the title of your proposal.

I added the javadoc.

### CODE REVIEW

File: src/main/java/javax/validation/valueextraction/ValueExtractor.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.validation.Path;
+
+/**
+ * Defines the logic used to extract the values to validate from an object of type {@code T}.
+ * <p>
+ * A {@link ValueExtractor} is tied to one specific type parameter of {@code T}. The {@link ExtractedValue} annotation
```
Comment (gunnarmorling) [inline src/main/java/javax/validation/valueextraction/ValueExtractor.java]: No need to link to itself. "A value extractor..."

File: src/main/java/javax/validation/valueextraction/ValueExtractor.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		/**
+		 * Receives the value extracted from an object.
+		 *
+		 * @param object the value to validate
+		 * @param nodeName the name of the node to be added to the {@link Path}
+		 */
+		void value(Object object, String nodeName);
+
+		/**
+		 * Receives the value extracted from an {@link Iterable} object that is not indexed (typically a {@link Set}).
```
Comment (gunnarmorling) [inline src/main/java/javax/validation/valueextraction/ValueExtractor.java]: I'd say "typically = e.g."

File: src/main/java/javax/validation/valueextraction/ValueExtractor.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import javax.validation.Path;
+
+/**
+ * Defines the logic used to extract the values to validate from an object of type {@code T}.
+ * <p>
+ * A {@link ValueExtractor} is tied to one specific type parameter of {@code T}. The {@link ExtractedValue} annotation
+ * is used to mark that type parameter.
+ * <p>
+ * The extracted values are passed to the corresponding method of the {@link ValueReceiver}.
```
Comment (gunnarmorling) [inline src/main/java/javax/validation/valueextraction/ValueExtractor.java]: It'd be good to give an example definition showing the usage of `@ExtractedValue` in the JavaDocs.

File: src/main/java/javax/validation/valueextraction/ValueExtractor.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	 */
+	interface ValueReceiver {
+
+		/**
+		 * Receives the value extracted from an object.
+		 *
+		 * @param object the value to validate
+		 * @param nodeName the name of the node to be added to the {@link Path}
+		 */
+		void value(Object object, String nodeName);
```
Comment (gunnarmorling) [inline src/main/java/javax/validation/valueextraction/ValueExtractor.java]: @emmanuelbernard suggested - and I had the same inclination before - to switch the parameter order: first the "meta data", i.e. node name/key/index, and then the value. That seems more natural to use.

**Comment by gsmet** [comment]
I fixed all your remarks.

> @gsmet Are you going to add @ConstraintsApplyTo, too (replacement for HV's @UnwrapValidatedValue)?

I think we should get this committed first. `@ConstraintsApplyTo` is not a simple replacement of the existing annotation and we don't have an implementation yet. It should be part of a follow-up commit.

**Comment by gunnarmorling** [comment]
> It should be part of a follow-up commit.

Ok.

**Comment by gunnarmorling** [comment]
Merged, thanks. I've pushed one more commit for some JavaDoc adjustments.

**Pull request merged**